ceno_zkvm/examples/fibonacci_elf.rs

```diff
@@ -0,0 +1,353 @@
+
+    let config = Rv32imConfig::<E>::construct_circuits(&mut zkvm_cs);
+    let mmu_config = MmuConfig::<E>::construct_circuits(&mut zkvm_cs);
+    let dummy_config = DummyExtraConfig::<E>::construct_circuits(&mut zkvm_cs);
+    let prog_config = zkvm_cs.register_table_circuit::<ExampleProgramTableCircuit<E>>();
+    zkvm_cs.register_global_state::<GlobalState>();
+
+    let mut zkvm_fixed_traces = ZKVMFixedTraces::default();
+
+    zkvm_fixed_traces.register_table_circuit::<ExampleProgramTableCircuit<E>>(
+        &zkvm_cs,
+        &prog_config,
+        vm.program(),
+    );
+
+    let mem_init = {
+        let program_addrs = vm
+            .program()
+            .image
+            .iter()
+            .map(|(addr, value)| MemInitRecord {
+                addr: *addr,
+                value: *value,
+            });
+
+        let stack_addrs = (1..=STACK_SIZE)
+            .map(|i| STACK_TOP - i * WORD_SIZE as u32)
+            .map(|addr| MemInitRecord { addr, value: 0 });
+
+        let mem_init = chain!(program_addrs, stack_addrs).collect_vec();
+
+        mem_padder.padded_sorted(MmuConfig::<E>::static_mem_len(), mem_init)
+    };
+
+    // IO is not used in this program, but it must have a particular size at the moment.
+    let io_init = mem_padder.padded_sorted(MmuConfig::<E>::public_io_len(), vec![]);
+
+    let reg_init = MmuConfig::<E>::initial_registers();
+    config.generate_fixed_traces(&zkvm_cs, &mut zkvm_fixed_traces);
+    mmu_config.generate_fixed_traces(
+        &zkvm_cs,
+        &mut zkvm_fixed_traces,
+        &reg_init,
+        &mem_init,
+        &io_init.iter().map(|rec| rec.addr).collect_vec(),
+    );
+    dummy_config.generate_fixed_traces(&zkvm_cs, &mut zkvm_fixed_traces);
+
+    let pk = zkvm_cs
+        .clone()
+        .key_gen::<Pcs>(pp.clone(), vp.clone(), zkvm_fixed_traces.clone())
+        .expect("keygen failed");
+    let vk = pk.get_vk();
+
+    // proving
+    let e2e_start = Instant::now();
+    let prover = ZKVMProver::new(pk);
+    let verifier = ZKVMVerifier::new(vk);
+
+    let all_records = vm
+        .iter_until_halt()
+        .take(args.max_steps.unwrap_or(usize::MAX))
+        .collect::<Result<Vec<StepRecord>, _>>()
+        .expect("vm exec failed");
+
+    tracing::info!("Proving {} execution steps", all_records.len());
+    for (i, step) in enumerate(&all_records).rev().take(5).rev() {
+        tracing::trace!("Step {i}: {:?} - {:?}\n", step.insn().codes().kind, step);
+    }
+
+    // Find the exit code from the HALT step, if halting at all.
+    let exit_code = all_records
+        .iter()
+        .rev()
+        .find(|record| {
+            record.insn().codes().kind == EANY
+                && record.rs1().unwrap().value == CENO_PLATFORM.ecall_halt()
+        })
+        .and_then(|halt_record| halt_record.rs2())
+        .map(|rs2| rs2.value);
+
+    let final_access = vm.tracer().final_accesses();
+    let end_cycle: u32 = vm.tracer().cycle().try_into().unwrap();
```

Comment:
If `PublicValues` needs a `u32` for end_cycle, then `fn cycle` should probably return a `u32`?